src/dbn_struct.jl

```diff
@@ -0,0 +1,222 @@
+           ),
+           (s = (),
+            a = (),
+            sp = map(DBNVar, (:s, :a)),
+            o = map(DBNVar, (:s, :a, :sp)),
+            r = map(DBNVar, (:s, :a, :sp, :o)),
+           )
+          )
+end
+
+"""
+    DBNStructure(::Type{M}) where M <: Union{MDP, POMDP}
+
+Trait of an MDP/POMDP type for describing the structure of the dynamic Baysian network.
+
+# Example
+
+    struct MyMDP <: MDP{Int, Int} end
+    POMDPs.gen(::MyMDP, s, a, rng) = (sp=s+a+rand(rng, [1,2,3]), r=s^2)
+
+    # make a new node, delta_s, that is deterministically equal to sp - s
+    function POMDPs.DBNStructure(::Type{MyMDP})
+        dbn = mdp_dbn()
+        return add_node(dbn, :delta_s, FunctionDBNNode((m,s,sp)->sp-s), (:s, :sp))
+    end
+
+    gen(DBNOut(:delta_s), MyMDP(), 1, 1, Random.GLOBAL_RNG)
+"""
+function DBNStructure end
+
+DBNStructure(::Type{M}) where M <: MDP = mdp_dbn()
+DBNStructure(::Type{M}) where M <: POMDP = pomdp_dbn()
+
+DBNStructure(m) = DBNStructure(typeof(m))
+
+struct InputDBNNode end # this does nothing for now
+
+"""
+DBN node defined by a function that maps the model and values from the parent nodes to a distribution
+
+# Example
+    DistributionDBNNode((m, s, a)->POMDPModelTools.Deterministic(s+a))    
+"""
+struct DistributionDBNNode{F}
+    dist_func::F
+end
+
+@generated function gen(n::DistributionDBNNode, m, args...)
+    # apparently needs to be @generated for type stability
+    argexpr = (:(args[$i]) for i in 1:length(args)-1)
+    quote
+        rand(last(args), n.dist_func(m, $(argexpr...)))
+    end
+end
+
+function implemented(g::typeof(gen), n::DistributionDBNNode, M, Deps, RNG)
+    return implemented(n.dist_func, Tuple{M, Deps.parameters...})
+end
+
+
+"""
+DBN node defined by a function that determinisitically maps the model and values from the parent nodes to a new value.
+
+# Example
+    FunctionDBNNode((m, s, a)->s+a)
+"""
+struct FunctionDBNNode{F}
+    f::F
+end
+
+@generated function gen(n::FunctionDBNNode, m, args...)
+    # apparently this needs to be @generated for type stability
+    argexpr = (:(args[$i]) for i in 1:length(args)-1)
+    quote
+        n.f(m, $(argexpr...))
+    end
+end
+
+function implemented(g::typeof(gen), n::FunctionDBNNode, M, Deps, RNG)
+    return implemented(n.f, Tuple{M, Deps.parameters...})
+end
+
+"""
+DBN node that always takes a deterministic constant value.
+"""
+struct ConstantDBNNode{T}
+    val::T
+end
+
+gen(n::ConstantDBNNode, args...) = n.val
+implemented(g::typeof(gen), n::ConstantDBNNode, M, Deps, RNG) = true
+
+"""
+DBN node that can only have a generative model; `gen(::DBNVar{:x}, ...)` must be implemented for a node of this type.
+"""
+struct GenDBNNode end
+
+gen(::GenDBNNode, args...) = error("No `gen(::DBNVar, ...)` method implemented for a GenDBNNode (see stack trace for name)")
+implemented(g::typeof(gen), GenDBNNode, M, Deps, RNG) = false
+
+function sorted_deppairs(dbn::Type{D}, symbols) where D <: DBNStructure
```

Comment:
No data type for `symbols` ?

Reply:
yeah, it could be any collection of symbols, so I don't want to constrain the type - I put a note about it in the docstring above